src/node/commit.rs

```diff
@@ -0,0 +1,191 @@
+    // Two nodes are equal if they both have IMRs and those IMRs are equal.
+    fn eq(&self, other: &Self) -> bool {
+        self.imr
+            .zip(other.imr)
+            .map(|(a, b)| a == b)
+            .unwrap_or(false)
+    }
+}
+
+impl<J: Jet> CommitData<J> {
+    /// Accessor for the node's arrow
+    pub fn arrow(&self) -> &FinalArrow {
+        &self.arrow
+    }
+
+    /// Helper function to compute a cached first-pass IMR
+    fn first_pass_imr(inner: Inner<&Arc<Self>, J, NoWitness>) -> Option<FirstPassImr> {
+        match inner {
+            Inner::Iden => Some(FirstPassImr::iden()),
+            Inner::Unit => Some(FirstPassImr::unit()),
+            Inner::InjL(child) => child.first_pass_imr.map(FirstPassImr::injl),
+            Inner::InjR(child) => child.first_pass_imr.map(FirstPassImr::injr),
+            Inner::Take(child) => child.first_pass_imr.map(FirstPassImr::take),
+            Inner::Drop(child) => child.first_pass_imr.map(FirstPassImr::drop),
+            Inner::Comp(left, right) => left
+                .first_pass_imr
+                .zip(right.first_pass_imr)
+                .map(|(a, b)| FirstPassImr::comp(a, b)),
+            Inner::Case(left, right) => left
+                .first_pass_imr
+                .zip(right.first_pass_imr)
+                .map(|(a, b)| FirstPassImr::case(a, b)),
+            Inner::AssertL(left, r_cmr) => left
+                .first_pass_imr
+                .map(|l_imr| FirstPassImr::case(l_imr, r_cmr.into())),
+            Inner::AssertR(l_cmr, right) => right
+                .first_pass_imr
+                .map(|r_imr| FirstPassImr::case(l_cmr.into(), r_imr)),
+            Inner::Pair(left, right) => left
+                .first_pass_imr
+                .zip(right.first_pass_imr)
+                .map(|(a, b)| FirstPassImr::pair(a, b)),
+            Inner::Disconnect(..) => None,
+            Inner::Witness(..) => None,
+            Inner::Fail(entropy) => Some(FirstPassImr::fail(entropy)),
+            Inner::Jet(jet) => Some(FirstPassImr::jet(jet)),
+            Inner::Word(ref val) => Some(FirstPassImr::const_word(val)),
+        }
+    }
+
+    pub fn new(
+        arrow: &Arrow,
+        inner: Inner<&Arc<Self>, J, NoWitness>,
+    ) -> Result<Self, types::Error> {
+        let final_arrow = arrow.finalize()?;
+        let first_pass_imr = Self::first_pass_imr(inner);
+        Ok(CommitData {
+            first_pass_imr,
+            imr: first_pass_imr.map(|imr| Imr::compute_pass2(imr, &final_arrow)),
+            arrow: final_arrow,
+            phantom: PhantomData,
+        })
+    }
+
+    pub fn new_unfinal(arrow: FinalArrow, inner: Inner<&Arc<Self>, J, NoWitness>) -> Self {
+        let first_pass_imr = Self::first_pass_imr(inner);
+        CommitData {
+            first_pass_imr,
+            imr: first_pass_imr.map(|imr| Imr::compute_pass2(imr, &arrow)),
+            arrow,
+            phantom: PhantomData,
+        }
+    }
```

Comment:
b3bce4e36dc504598954839cb0c9ab3556a2725e: Kind of confusing how `CommitNode::new_unfinal` gets the `FinalArrow`. I get it, it is used in `RedeemNode::unfinalize`, but the name doesn't make sense in the context of `CommitNode`. Maybe call it `CommitNode::new_final` and make `CommitNode::new` call `CommitNode::new_final`?